internal/wasm/sdk/opa/opa_bench_test.go

```diff
@@ -0,0 +1,156 @@
+	"fmt"
+	"testing"
+
+	"github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm"
+	"github.com/open-policy-agent/opa/internal/wasm/sdk/opa"
+	"github.com/open-policy-agent/opa/util/test"
+)
+
+func BenchmarkWASMArrayIteration(b *testing.B) {
+	sizes := []int{10, 100, 1000, 10000}
+	for _, n := range sizes {
+		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			benchmarkIteration(b, n, test.ArrayIterationBenchmarkModule(n))
+		})
+	}
+}
+
+func BenchmarkWASMSetIteration(b *testing.B) {
+	sizes := []int{10, 100, 1000, 10000}
+	for _, n := range sizes {
+		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			benchmarkIteration(b, n, test.SetIterationBenchmarkModule(n))
+		})
+	}
+}
+
+func BenchmarkWASMObjectIteration(b *testing.B) {
+	sizes := []int{10, 100, 1000, 10000}
+	for _, n := range sizes {
+		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			benchmarkIteration(b, n, test.ObjectIterationBenchmarkModule(n))
+		})
+	}
+}
+
+var r *opa.Result
+
+func benchmarkIteration(b *testing.B, n int, module string) {
+	query := "data.test.main = x"
+	policy := compileRegoToWasm(module, query, false)
+	// _ = ioutil.WriteFile(fmt.Sprintf("iteration_%d.wasm", n), policy, 0644)
+
+	instance, err := opa.New().
+		WithPolicyBytes(policy).
+		WithMemoryLimits(2*wasm.PageSize, 47*wasm.PageSize).
+		WithPoolSize(1).
+		Init()
+	if err != nil {
+		b.Fatalf("init sdk: %v", err)
+	}
+
+	b.ResetTimer()
+	ctx := context.Background()
+	var input interface{} = make(map[string]interface{})
+
+	for i := 0; i < b.N; i++ {
+		r, err = instance.Eval(ctx, opa.EvalOpts{Input: &input})
+		if err != nil {
+			b.Fatalf("Unexpected query error: %v", err)
+		}
+		if string(r.Result) != `{{"x":true}}` {
+			b.Errorf("unexpected result: %s", string(r.Result))
+		}
+		// b.ReportMetric(float64(r.MemUsed), "bytes/op")
+	}
+}
+
+func BenchmarkWASMLargeJSON(b *testing.B) {
+	for _, kv := range []struct{ key, val int }{
+		{10, 10},
+		{10, 100},
+		{10, 1000},
+		{10, 10000},
+		{100, 100},
+		{100, 1000},
+	} {
+		b.Run(fmt.Sprintf("%dx%d", kv.key, kv.val), func(b *testing.B) {
+			ctx := context.Background()
+			data := test.GenerateJSONBenchmarkData(kv.key, kv.val)
+
+			// Read data.values N times inside query.
+			query := "data.keys[_] = x; data.values = y"
+			policy := compileRegoToWasm("", query, false)
+
+			instance, err := opa.New().
+				WithPolicyBytes(policy).
+				WithDataJSON(data).
+				WithMemoryLimits(200*wasm.PageSize, 600*wasm.PageSize). // This is rather much
+				WithPoolSize(1).
+				Init()
+			if err != nil {
+				b.Fatalf("init sdk: %v", err)
+			}
+
+			b.ResetTimer()
+			var input interface{} = make(map[string]interface{})
+
+			for i := 0; i < b.N; i++ {
+				r, err = instance.Eval(ctx, opa.EvalOpts{Input: &input})
+				if err != nil {
+					b.Fatalf("Unexpected query error: %v", err)
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkWASMVirtualDocs(b *testing.B) {
+	for _, kv := range []struct{ total, hit int }{
+		{1, 1},
+		{10, 1},
+		{100, 1},
+		{1000, 1},
```

Comment:
I've been adding a few data points in between 100 and 1000 to glimpse at the line graph:

<img width="731" alt="Screenshot 2021-01-27 at 17 28 42" src="https://user-images.githubusercontent.com/870638/106021666-507b1700-60c5-11eb-9f15-eed729ce6e91.png">

```
goos: darwin
goarch: amd64
pkg: github.com/open-policy-agent/opa/internal/wasm/sdk/opa
BenchmarkWASMVirtualDocs/total=1/hit=1-16         	   60504	     20041 ns/op
BenchmarkWASMVirtualDocs/total=10/hit=1-16        	   51378	     21870 ns/op
BenchmarkWASMVirtualDocs/total=100/hit=1-16       	   33804	     36900 ns/op
BenchmarkWASMVirtualDocs/total=200/hit=1-16       	   20919	     55119 ns/op
BenchmarkWASMVirtualDocs/total=300/hit=1-16       	   15438	     76811 ns/op
BenchmarkWASMVirtualDocs/total=500/hit=1-16       	   10000	    116501 ns/op
BenchmarkWASMVirtualDocs/total=750/hit=1-16       	    7716	    159633 ns/op
BenchmarkWASMVirtualDocs/total=1000/hit=1-16      	    5395	    195809 ns/op
PASS
ok  	github.com/open-policy-agent/opa/internal/wasm/sdk/opa	23.376s
```

Reply:
https://github.com/cep21/benchdraw was the first thing I've found that seemed like it required very little effort to use 😄